modules/apps/friendly-url/friendly-url-service/src/main/java/com/liferay/friendly/url/internal/servlet/FriendlyURLServlet.java

```diff
@@ -343,7 +343,28 @@ public Redirect getRedirect(
 			Layout redirectLayout = null;
 
 			if (layoutFriendlyURL == null) {
-				redirectLayout = defaultLayout;
+				if (exception instanceof LayoutPermissionException) {
+					List<Layout> layouts = layoutLocalService.getLayouts(
+						group.getGroupId(), _private,
+						LayoutConstants.DEFAULT_PARENT_LAYOUT_ID);
+
+					PermissionChecker permissionChecker =
+						PermissionCheckerFactoryUtil.create(
+							_getUser(httpServletRequest));
+
+					for (Layout layout : layouts) {
+						if (LayoutPermissionUtil.contains(
+								permissionChecker, layout, ActionKeys.VIEW)) {
+
+							redirectLayout = layout;
+
+							break;
+						}
+					}
+				}
```

Comment:
@ChrisKian Is there any reason for not using the equivalent LayoutService#getLayouts() method? It seems to be doing the same thing as this chunk of code, so by using that instead of the local version this would be much simpler.

Reply:
@adolfopa, the current usage is used as we only need to return the first viewable layout whereas LayoutService#getLayouts() will return all viewable layouts.

Reply:
I've added the changes @adolfopa.  My only concern is the filtering of layouts, as it relates to similar methods.

[Currently in LayoutServiceImpl](https://github.com/liferay/liferay-portal/blob/master/portal-impl/src/com/liferay/portal/service/impl/LayoutServiceImpl.java#L736-L746), when using a `getLayouts` method with a `start` and `end` argument, we first retrieve all layouts, delimit by `start`/`end`, then we filter based on permissions.  This won't necessarily work for our use-case, so I had to switch the order of when we filter and when we delimit.  Is this acceptable for this new method only, should we do it for all methods, or is my approach incorrect completely?  Thanks!